### ISSUE TITLE
Architecture high-level overview

### ISSUE THREAD

**Issue by dszlachta** [TrueBlocks/trueblocks-key#1]
Suggestion of TrueBlocks-QN project architecture.

### Cost-effective data delivery
We don't need to support all `chifra` features. We only need to serve "give me all appearances of address X" requests. 

Because the index is incremental (each new block can add data, but never remove or change it), we can prepare the data in needed format and store in cloud. This is called object storage and is one of the cheapest features of cloud solutions. Here's how it can be achieved:

1. We run scraper. It can run on any machine (e.g. the one we already use for running tests). Scraper produces chunk files
2. On the same machine a simple Go program (`converter`) is watching index directory. It extracts appearances from index and convert them into JSON files. 
3. The file structure that `converter` creates is `address/appearances.json` (makes data accessible without any dynamic logic)
4. The output directory of `converter` is rsynced with cloud object storage. Cloud storage is updated automatically when there's a change in local directory.

The above makes data available in cloud and possible to query it. But we need to make sure only QN users have access to it.

### User access

When a new user subscribes for TrueBlocks on Marketplace, QN will send us a request with this user's ID (ID is permanent). We need to store it. ID is also sent with every request that the user makes. 

We need another Go program (`handler`) that would:
1. save new user's ID to a database
2. remove user ID when they cancel subscription
3. for each data request check if user ID is in our database
4. serve data (if user ID is in the database)

Luckily for us, we don't need to stand up cloud instance (expensive) to run `handler`. We can use serverless cloud functions - meant to run simple code and way cheaper because operating system and infrastructure is abstracted away. 

### Cloud architecture

![Diagram bez tytułu drawio](https://github.com/TrueBlocks/trueblocks-qn/assets/2346536/803d25a3-d11d-4144-a72e-c72dc765ac4e)

### Providers we can use

**AWS**
- S3 for object storage 
- AWS Lambda for serverless functions
- DynamoDB for cheap user database (we don't need SQL)
AWS is the cheapest option, but configuring it is a nightmare. We can still do it, though.

**DigitalOcean**
- Spaces for object storage
- Functions for serverless functions
- Postgres for user database
DigitalOcean is bit more expensive, but it's easy to configure.

When the code is working we can measure memory consumption and execution time and calculate costs for each provider. I'm pretty sure we're talking about tens of $, maybe around $100 in the worst case. 

Basically to make it as cheap as possible we should keep away from running cloud instances and instead leverage object storage, cheap databases and make sure our serverless functions have high performance.
